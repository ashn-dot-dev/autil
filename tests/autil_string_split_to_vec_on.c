#define AUTIL_IMPLEMENTATION
#include "../autil.h"
#include "test.h"

static void
cleanup(struct autil_string* string, struct autil_vec* split)
{
    autil_string_del(string);
    for (size_t i = 0; i < autil_vec_count(split); ++i) {
        autil_string_del(
            AUTIL_DEREF_PTR(struct autil_string*, autil_vec_ref(split, i)));
    }
}

static void
dump(struct autil_string const* string, struct autil_vec const* split)
{
    printf("\"%s\" =>", autil_string_start(string));
    for (size_t i = 0; i < autil_vec_count(split); ++i) {
        struct autil_string const* const* ps = autil_vec_ref_const(split, i);
        printf(" \"%s\"", autil_string_start(*ps));
    }
    fputc('\n', stdout);
}

int
main(void)
{
    struct autil_string* string = NULL;
    struct autil_vec* const split = autil_vec_new(sizeof(struct autil_string*));

    // Main tests mostly testing core functionality and cstr splits.
    string = autil_string_new_cstr("");
    autil_string_split_to_vec_on_cstr(string, "", split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("A");
    autil_string_split_to_vec_on_cstr(string, "", split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("A");
    autil_string_split_to_vec_on_cstr(string, "A", split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("AA");
    autil_string_split_to_vec_on_cstr(string, "A", split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("A A");
    autil_string_split_to_vec_on_cstr(string, " ", split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("ABCBB");
    autil_string_split_to_vec_on_cstr(string, "B", split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("ABC");
    autil_string_split_to_vec_on_cstr(string, "D", split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("ABC");
    autil_string_split_to_vec_on_cstr(string, "FOOBAR", split);
    dump(string, split);
    cleanup(string, split);

    fputc('\n', stdout);
    // Some vstr tests.
    string = autil_string_new_cstr("ABCBB");
    autil_string_split_to_vec_on_vstr(
        string, AUTIL_VSTR_LOCAL_PTR_STR_LITERAL("B"), split);
    dump(string, split);
    cleanup(string, split);

    string = autil_string_new_cstr("ABC");
    autil_string_split_to_vec_on_vstr(
        string, AUTIL_VSTR_LOCAL_PTR_STR_LITERAL("FOOBAR"), split);
    dump(string, split);
    cleanup(string, split);

    autil_vec_del(split);
}
