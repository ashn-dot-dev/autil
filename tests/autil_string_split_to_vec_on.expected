"" => ""
"A" => "A"
"A" => "" ""
"AA" => "" "" ""
"A A" => "A" "A"
"ABCBB" => "A" "C" "" ""
"ABC" => "ABC"
"ABC" => "ABC"

"ABCBB" => "A" "C" "" ""
"ABC" => "ABC"
